snorkel/candidates.py

```diff
@@ -1,24 +1,70 @@
-import sys
-from collections import defaultdict
 from itertools import chain
-from time import time
 from multiprocessing import Process, Queue, JoinableQueue
+from snorkel import SnorkelBase, SnorkelSession
+from parser import Context
+from sqlalchemy import Column, String, Integer, ForeignKey
+from sqlalchemy.orm import relationship, backref
+from sqlalchemy.types import PickleType
 from Queue import Empty
 
 
-class Candidate(object):
-    """A candidate object, **uniquely identified by its id**"""
-    def __init__(self, id):
-        self.id = id
-    
-    def __eq__(self, other):
-        try:
-            return self.id == other.id
-        except:
-            raise NotImplementedError()
-    
-    def __hash__(self):
-        return hash(self.id)
+QUEUE_COLLECT_TIMEOUT = 5
+
+
+class Candidates(SnorkelBase):
+    """A named collection of Candidate objects."""
+    __tablename__ = 'candidates'
+    id = Column(String, primary_key=True)
+
+    def __repr__(self):
+        return "Candidates" + str((self.id,))
+
+    def __iter__(self):
+        """Default iterator is over Candidate objects"""
+        for candidate in self.candidates:
+            yield candidate
+
+    def get_candidates(self):
+        return self.candidates
+
+    def get_candidate(self, candidate_id):
+        """Retrieve a Candidate by id"""
+        session = SnorkelSession.object_session(self)
```

Comment:
Good question. In that case the object's session is None. A possible fix is

```python
def get_candidate(self, candidate_id):
        session = SnorkelSession.object_session(self)
        if session is None:
            session = SnorkelSession()
            session.add(self)
            cand = session.query(Candidate).filter(Candidate.id == candidate_id).one()
            session.close()
            return cand
        else:
            return session.query(Candidate).filter(Candidate.id == candidate_id).one()
```

but this is running off the rails. See http://docs.sqlalchemy.org/en/latest/orm/session_basics.html#session-frequently-asked-questions

I think the issue is that canned queries without some externally supplied session in which to execute them is not the SQLAlchemy way AFAIK. So better fixes are probably either (1) take a session as an argument or (2) get rid of these methods and use session queries directly for specific candidates, documents, etc.